phpBB/phpbb/module/module_manager.php

```
@@ -0,0 +1,552 @@
+					AND ' . $this->db->sql_in_set('module_id', $moved_ids, true);
+			$this->db->sql_query($sql);
+
+			// Resync the righthand side of the tree
+			$sql = 'UPDATE ' . $this->modules_table . "
+				SET left_id = left_id + $diff, right_id = right_id + $diff
+				WHERE module_class = '" . $this->db->sql_escape($module_class) . "'
+					AND left_id > " . (int) $to_data['right_id'] . '
+					AND ' . $this->db->sql_in_set('module_id', $moved_ids, true);
+			$this->db->sql_query($sql);
+
+			// Resync moved branch
+			$to_data['right_id'] += $diff;
+			if ($to_data['right_id'] > $from_data['right_id'])
+			{
+				$diff = '+ ' . ($to_data['right_id'] - $from_data['right_id'] - 1);
+			}
+			else
+			{
+				$diff = '- ' . abs($to_data['right_id'] - $from_data['right_id'] - 1);
+			}
+		}
+		else
+		{
+			$sql = 'SELECT MAX(right_id) AS right_id
+				FROM ' . $this->modules_table . "
+				WHERE module_class = '" . $this->db->sql_escape($module_class) . "'
+					AND " . $this->db->sql_in_set('module_id', $moved_ids, true);
+			$result = $this->db->sql_query($sql);
+			$row = $this->db->sql_fetchrow($result);
+			$this->db->sql_freeresult($result);
+
+			$diff = '+ ' . (int) ($row['right_id'] - $from_data['left_id'] + 1);
+		}
+
+		$sql = 'UPDATE ' . $this->modules_table . "
+			SET left_id = left_id $diff, right_id = right_id $diff
+			WHERE module_class = '" . $this->db->sql_escape($module_class) . "'
+				AND " . $this->db->sql_in_set('module_id', $moved_ids);
+		$this->db->sql_query($sql);
+	}
+
+	/**
+	 * Remove module from tree
+	 *
+	 * @param int		$module_id		ID of the module to delete
+	 * @param string	$module_class	Class of the module (acp, ucp, mcp etc...)
+	 *
+	 * @return array
+	 */
+	public function delete_module($module_id, $module_class)
+	{
+		$row = $this->get_module_row($module_id, $module_class);
+
+		$branch = $this->get_module_branch($module_id, $module_class, 'children', false);
+
+		if (sizeof($branch))
+		{
+			return array('CANNOT_REMOVE_MODULE');
```

should we throw an exception instead?